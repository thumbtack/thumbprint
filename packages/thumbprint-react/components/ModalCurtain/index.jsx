import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import NoScroll from './components/no-scroll.jsx';

import useCloseOnEscape from '../../utils/use-close-on-escape';
import useFocusTrap from '../../utils/use-focus-trap';
import ConditionalPortal from '../../utils/ConditionalPortal';

import styles from './index.module.scss';

const propTypes = {
    /**
     * Content that appears on top of the curtain. `children` is a [render
     * prop](https://reactjs.org/docs/render-props.html) and expects a function. The function
     * provided receives an object two properties:
     *
     * * `curtainClassName`: Handles positioning, z-index, opacity, overflow, and visibility.
     * Should be added to the `className` of the outermost element within the `children`.
     * * `curtainOnClick`: You can optionally add this prop to the `onClick` of an element within
     * `children`. This is typically used on to close a ModalCurtain when clicking on a dark
     * backdrop.
     *
     * The examples on this page include code samples that demonstrate real use of these props.
     */
    children: PropTypes.func,
    /**
     * The four states that a modal can be in.
     *
     * * `exited` – Modal is fully closed
     * * `entering` – Modal is opening
     * * `entered` – Modal has fully open
     * * `exiting` – Modal is closing
     *
     * Modals that do not have transitions will only use the `entered` and `exited` stages.
     */
    stage: PropTypes.oneOf(['entering', 'entered', 'exiting', 'exited']),
    /**
     * Accessibility title used to describe the content of the modal to screen readers.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Function that fires to close the modal.
     */
    onCloseClick: PropTypes.func.isRequired,
    /**
     * Determines if the modal should close when pressing the escape key.
     */
    shouldCloseOnEscape: PropTypes.bool,
};

const defaultProps = {
    children: undefined,
    accessibilityLabel: 'Modal',
    stage: 'exited',
    shouldCloseOnEscape: true,
};

export default function ModalCurtain({
    stage,
    shouldCloseOnEscape,
    accessibilityLabel,
    onCloseClick,
    children,
}) {
    const [isClient, setIsClient] = useState(false);
    const [wrapperEl, setWrapperEl] = useState(null);

    useEffect(() => {
        setIsClient(true);
    }, []);

    const isEnteringOrEntered = stage === 'entering' || stage === 'entered';
    const shouldBindEscListener = isClient && shouldCloseOnEscape && isEnteringOrEntered;
    const shouldTrapFocus = isClient && wrapperEl && stage === 'entered';
    const shouldDisableScrolling = isEnteringOrEntered;

    useCloseOnEscape(onCloseClick, shouldBindEscListener);
    useFocusTrap(wrapperEl, shouldTrapFocus);

    return (
        <ConditionalPortal shouldDisplace>
            {/* Use tabIndex="-1" to allow programmatic focus (as initialFocus node for focus-trap)
            but not be tabbable by user. */}
            <div
                role="dialog"
                aria-label={accessibilityLabel}
                tabIndex="-1"
                ref={element => {
                    setWrapperEl(element);
                }}
            >
                {shouldDisableScrolling && <NoScroll />}

                {/* This component uses the render prop pattern. `children` expects a function and
                receives an object that contains `curtainOnClick` and `curtainClassName`.
                While using those two properties is optional, they provide helpful functionality. */}
                {children &&
                    children({
                        curtainOnClick: event => {
                            // Ensures that the click event happened on the element that has the
                            // `onClick`. This prevents clicks deep within `children` from bubbling
                            // up and closing the ModalCurtain.
                            if (event.target === event.currentTarget) {
                                onCloseClick();
                            }
                        },
                        curtainClassName: classNames({
                            [styles.root]: true,
                            [styles.rootOpen]: isEnteringOrEntered,
                        }),
                    })}
            </div>
        </ConditionalPortal>
    );
}

ModalCurtain.propTypes = propTypes;
ModalCurtain.defaultProps = defaultProps;
