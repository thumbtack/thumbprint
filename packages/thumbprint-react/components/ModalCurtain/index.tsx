import React, { useState, useEffect } from 'react';
import classNames from 'classnames';

import NoScroll from './components/no-scroll';

import useCloseOnEscape from '../../utils/use-close-on-escape';
import useFocusTrap from '../../utils/use-focus-trap';
import ConditionalPortal from '../../utils/ConditionalPortal';

import styles from './index.module.scss';

interface PropTypes {
    /**
     * Content that appears on top of the curtain. `children` is a [render
     * prop](https://reactjs.org/docs/render-props.html) and expects a function. The function
     * provided receives an object two properties:
     *
     * * `curtainClassName`: Handles positioning, z-index, opacity, overflow, and visibility.
     * Should be added to the `className` of the outermost element within the `children`.
     * * `curtainOnClick`: You can optionally add this prop to the `onClick` of an element within
     * `children`. This is typically used on to close a ModalCurtain when clicking on a dark
     * backdrop.
     *
     * The examples on this page include code samples that demonstrate real use of these props.
     */
    children?: ({
        curtainClassName,
        curtainOnClick,
    }: {
        curtainClassName: string;
        curtainOnClick: (event: React.MouseEvent<HTMLElement>) => void;
    }) => JSX.Element;
    /**
     * The four states that a modal can be in.
     *
     * * `exited` – Modal is fully closed
     * * `entering` – Modal is opening
     * * `entered` – Modal has fully open
     * * `exiting` – Modal is closing
     *
     * Modals that do not have transitions will only use the `entered` and `exited` stages.
     */
    stage?: 'entering' | 'entered' | 'exiting' | 'exited' | null;
    /**
     * Accessibility title used to describe the content of the modal to screen readers.
     */
    accessibilityLabel?: string;
    /**
     * Function that fires to close the modal.
     */
    onCloseClick: () => void;
    /**
     * Determines if the modal should close when pressing the escape key.
     */
    shouldCloseOnEscape?: boolean;
    /**
     * The element that should be focused when the modal opens. If omitted, the entire container
     * element of the modal is focused.
     */
    initialFocus?: HTMLElement | null;
}

export default function ModalCurtain({
    stage = 'exited',
    shouldCloseOnEscape = true,
    accessibilityLabel = 'Modal',
    initialFocus: initialFocusProp,
    onCloseClick,
    children,
}: PropTypes): JSX.Element {
    const [isClient, setIsClient] = useState<boolean>(false);
    const [wrapperEl, setWrapperEl] = useState<HTMLDivElement | null>(null);

    useEffect(() => {
        setIsClient(true);
    }, []);

    const initialFocus = initialFocusProp || wrapperEl;

    const isEnteringOrEntered = stage === 'entering' || stage === 'entered';
    const shouldBindEscListener = isClient && shouldCloseOnEscape && isEnteringOrEntered;
    const shouldTrapFocus = isClient && !!initialFocus && stage === 'entered';
    const shouldDisableScrolling = isEnteringOrEntered;

    useCloseOnEscape(onCloseClick, shouldBindEscListener);
    useFocusTrap(wrapperEl, shouldTrapFocus, initialFocus);

    return (
        <ConditionalPortal shouldDisplace>
            {/* Use tabIndex="-1" to allow programmatic focus (as initialFocus node for focus-trap)
            but not be tabbable by user. */}
            <div
                role="dialog"
                aria-label={accessibilityLabel}
                tabIndex={-1}
                ref={(element): void => {
                    setWrapperEl(element);
                }}
            >
                {shouldDisableScrolling && <NoScroll />}

                {/* This component uses the render prop pattern. `children` expects a function and
                receives an object that contains `curtainOnClick` and `curtainClassName`.
                While using those two properties is optional, they provide helpful functionality. */}
                {children &&
                    children({
                        curtainOnClick: event => {
                            // Ensures that the click event happened on the element that has the
                            // `onClick`. This prevents clicks deep within `children` from bubbling
                            // up and closing the ModalCurtain.
                            if (event.target === event.currentTarget) {
                                onCloseClick();
                            }
                        },
                        curtainClassName: classNames({
                            [styles.root]: true,
                            [styles.rootOpen]: isEnteringOrEntered,
                        }),
                    })}
            </div>
        </ConditionalPortal>
    );
}
