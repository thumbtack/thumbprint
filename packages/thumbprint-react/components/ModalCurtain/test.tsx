import React, { useState } from 'react';
import { mount } from 'enzyme';
import ModalCurtain from './index';

const ESC_KEY = 27;
const ENTER_KEY = 13;
const BACKSPACE_KEY = 8;
const SPACE_KEY = 32;
const LOWERCASE_A_KEY = 65;

describe('ModalCurtain', () => {
    test('renders a basic curtain', () => {
        const wrapper = mount(<ModalCurtain onCloseClick={jest.fn} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders modal in `exited` stage', () => {
        const wrapper = mount(<ModalCurtain stage="exited" onCloseClick={jest.fn} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders modal in `entering` stage', () => {
        const wrapper = mount(<ModalCurtain stage="entering" onCloseClick={jest.fn} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders modal in `entered` stage', () => {
        const wrapper = mount(<ModalCurtain stage="entered" onCloseClick={jest.fn} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders modal in `exiting` stage', () => {
        const wrapper = mount(<ModalCurtain stage="exiting" onCloseClick={jest.fn} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders `accessibilityLabel`', () => {
        const wrapper = mount(
            <ModalCurtain accessibilityLabel="goosegoosegoose" onCloseClick={jest.fn} />,
        );
        expect(wrapper.html()).toContain('goosegoosegoose');
        expect(wrapper).toMatchSnapshot();
    });

    test('`children` receives `curtainOnClick` and `curtainClassName`', () => {
        let obj: {
            curtainOnClick?: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
            curtainClassName?: string;
        } = {};

        mount(
            <ModalCurtain onCloseClick={jest.fn}>
                {(args): JSX.Element => {
                    obj = args;
                    return <div />;
                }}
            </ModalCurtain>,
        );

        expect(obj.curtainOnClick).toBeTruthy();
        expect(obj.curtainClassName).toBeTruthy();
    });

    test("`children`'s `curtainOnClick` render prop calls `onCloseClick` prop when executed", () => {
        const onClick = jest.fn();

        const wrapper = mount(
            <ModalCurtain onCloseClick={onClick}>
                {({ curtainOnClick }): JSX.Element => (
                    <button data-test-id="goose" onClick={curtainOnClick} type="button" />
                )}
            </ModalCurtain>,
        );

        expect(onClick).toHaveBeenCalledTimes(0);

        wrapper.find('button[data-test-id="goose"]').simulate('click');

        expect(onClick).toHaveBeenCalledTimes(1);
    });

    test('renders text that is passed in', () => {
        const wrapper = mount(
            <div>
                <button type="button" />
                <ModalCurtain onCloseClick={jest.fn}>
                    {(): JSX.Element => <div>Goose</div>}
                </ModalCurtain>
            </div>,
        );

        expect(wrapper.find('ModalCurtain').text()).toEqual('Goose');
        expect(wrapper).toMatchSnapshot();
    });

    test('initially traps focus to the root dialog node', () => {
        jest.useFakeTimers();

        const onCloseClick = jest.fn();
        const wrapper = mount(
            <ModalCurtain stage="entered" onCloseClick={onCloseClick}>
                {(): JSX.Element => (
                    <>
                        <button type="button" />
                    </>
                )}
            </ModalCurtain>,
        );

        // Run setTimeouts() in focus-trap to completion
        jest.runAllTimers();

        const modalWrapper = wrapper.find('[role="dialog"]');
        expect(modalWrapper.is(':focus')).toBe(true);

        jest.useRealTimers();
    });

    test('initially focuses `initialFocus` element if provided', () => {
        jest.useFakeTimers();
        const onCloseClick = jest.fn();

        const Example = (): React.ReactElement => {
            const [inputRef, setInputRef] = useState<HTMLInputElement | null>(null);

            return (
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} initialFocus={inputRef}>
                    {(): JSX.Element => (
                        <div>
                            <button type="button" data-testid="button" />
                            <input
                                ref={(r): void => {
                                    setInputRef(r);
                                }}
                                data-testid="input"
                            />
                        </div>
                    )}
                </ModalCurtain>
            );
        };

        const wrapper = mount(<Example />);

        // Run setTimeouts() in focus-trap to completion
        jest.runAllTimers();

        // We don't want focus to be on the outside wrapper.
        const modalWrapper = wrapper.find('[role="dialog"]');
        expect(modalWrapper.is(':focus')).toBe(false);

        const button = wrapper.find('[data-testid="button"]');
        const input = wrapper.find('[data-testid="input"]');

        expect(button.is(':focus')).toBe(false);
        expect(input.is(':focus')).toBe(true);

        jest.useRealTimers();
    });

    test('`onCloseClick` is not called when non-ESC keys are pressed', () => {
        const onCloseClick = jest.fn();
        mount(<ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />);

        expect(onCloseClick).toHaveBeenCalledTimes(0);

        // NOTE: it seems that Enzyme does not respond to the standard properties on
        // `KeyboardEventInit`, either `code: 'Escape'` or `key: 'Escape'`. It only responds to
        // the deprecated property `keyCode`, which is not in the type defintion. As such we
        // have to cast the type here to prevent an error.
        document.dispatchEvent(
            new KeyboardEvent('keyup', {
                key: 'Enter',
                code: 'Enter',
                keyCode: ENTER_KEY,
            } as KeyboardEventInit),
        );
        document.dispatchEvent(
            new KeyboardEvent('keyup', {
                key: 'Backspace',
                code: 'Backspace',
                keyCode: BACKSPACE_KEY,
            } as KeyboardEventInit),
        );
        document.dispatchEvent(
            new KeyboardEvent('keyup', {
                key: 'Space',
                code: 'Space',
                keyCode: SPACE_KEY,
            } as KeyboardEventInit),
        );
        document.dispatchEvent(
            new KeyboardEvent('keyup', {
                key: 'a',
                code: 'a',
                keyCode: LOWERCASE_A_KEY,
            } as KeyboardEventInit),
        );

        expect(onCloseClick).toHaveBeenCalledTimes(0);
    });

    describe('closes modal on `ESC`', () => {
        test('when modal is open and `shouldCloseOnEscape` is `true`', () => {
            const onCloseClick = jest.fn();
            mount(<ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />);

            expect(onCloseClick).toHaveBeenCalledTimes(0);

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);
        });

        test('when modal is open and `shouldCloseOnEscape` are `true`, become closed and `false`, then become open and `true` again', () => {
            const onCloseClick = jest.fn();
            const wrapper = mount(
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />,
            );

            expect(onCloseClick).toHaveBeenCalledTimes(0);

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ shouldCloseOnEscape: false, stage: 'exited' });
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ shouldCloseOnEscape: true, stage: 'exited' });
            document.dispatchEvent(event);
            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ shouldCloseOnEscape: false, stage: 'entered' });
            document.dispatchEvent(event);
            expect(onCloseClick).toHaveBeenCalledTimes(1);

            // Both are now true, so the event listener should be reactiviated.
            wrapper.setProps({ shouldCloseOnEscape: true, stage: 'entered' });
            document.dispatchEvent(event);
            expect(onCloseClick).toHaveBeenCalledTimes(2);
        });
    });

    describe('does not close modal on `ESC`', () => {
        test('when modal is closed and `shouldCloseOnEscape` is `true`', () => {
            const onCloseClick = jest.fn();
            mount(<ModalCurtain stage="exited" onCloseClick={onCloseClick} shouldCloseOnEscape />);

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(0);
        });

        test('when modal is closed and `shouldCloseOnEscape` is `false`', () => {
            const onCloseClick = jest.fn();
            mount(
                <ModalCurtain
                    stage="exited"
                    onCloseClick={onCloseClick}
                    shouldCloseOnEscape={false}
                />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(0);
        });

        test('when modal is open and `shouldCloseOnEscape` is `false`', () => {
            const onCloseClick = jest.fn();
            mount(
                <ModalCurtain
                    stage="entered"
                    onCloseClick={onCloseClick}
                    shouldCloseOnEscape={false}
                />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(0);
        });

        test('when modal is open then changes to closed', () => {
            const onCloseClick = jest.fn();
            const wrapper = mount(
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ stage: 'exited' });

            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);
        });

        test('when `shouldCloseOnEscape` changes to `false` while mounted', () => {
            const onCloseClick = jest.fn();
            const wrapper = mount(
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ shouldCloseOnEscape: false });

            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);
        });

        test('when `shouldCloseOnEscape` changes to `false` and the modal close while mounted', () => {
            const onCloseClick = jest.fn();
            const wrapper = mount(
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.setProps({ shouldCloseOnEscape: false, stage: 'exited' });

            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);
        });

        test('when component is unmounted', () => {
            const onCloseClick = jest.fn();
            const wrapper = mount(
                <ModalCurtain stage="entered" onCloseClick={onCloseClick} shouldCloseOnEscape />,
            );

            const event = new KeyboardEvent('keyup', {
                key: 'Escape',
                code: 'Escape',
                keyCode: ESC_KEY,
            } as KeyboardEventInit);
            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);

            wrapper.unmount();

            document.dispatchEvent(event);

            expect(onCloseClick).toHaveBeenCalledTimes(1);
        });
    });
});
