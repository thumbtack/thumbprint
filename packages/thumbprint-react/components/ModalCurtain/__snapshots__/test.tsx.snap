// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalCurtain renders \`accessibilityLabel\` 1`] = `
<ModalCurtain
  accessibilityLabel="goosegoosegoose"
  onCloseClick={[Function]}
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="goosegoosegoose"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="goosegoosegoose"
        role="dialog"
        tabIndex={-1}
      />
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders a basic curtain 1`] = `
<ModalCurtain
  onCloseClick={[Function]}
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="Modal"
        role="dialog"
        tabIndex={-1}
      />
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders modal in \`entered\` stage 1`] = `
<ModalCurtain
  onCloseClick={[Function]}
  stage="entered"
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="Modal"
        role="dialog"
        tabIndex={-1}
      >
        <NoScroll />
      </div>
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders modal in \`entering\` stage 1`] = `
<ModalCurtain
  onCloseClick={[Function]}
  stage="entering"
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="Modal"
        role="dialog"
        tabIndex={-1}
      >
        <NoScroll />
      </div>
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders modal in \`exited\` stage 1`] = `
<ModalCurtain
  onCloseClick={[Function]}
  stage="exited"
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="Modal"
        role="dialog"
        tabIndex={-1}
      />
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders modal in \`exiting\` stage 1`] = `
<ModalCurtain
  onCloseClick={[Function]}
  stage="exiting"
>
  <ConditionalPortal
    shouldDisplace={true}
  >
    <Portal
      containerInfo={
        <body>
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
          <div
            aria-label="Modal"
            role="dialog"
            tabindex="-1"
          />
        </body>
      }
    >
      <div
        aria-label="Modal"
        role="dialog"
        tabIndex={-1}
      />
    </Portal>
  </ConditionalPortal>
</ModalCurtain>
`;

exports[`ModalCurtain renders text that is passed in 1`] = `
<div>
  <button
    type="button"
  />
  <ModalCurtain
    onCloseClick={[Function]}
  >
    <ConditionalPortal
      shouldDisplace={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="goosegoosegoose"
              role="dialog"
              tabindex="-1"
            />
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            >
              <div />
            </div>
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            >
              <button
                data-test-id="goose"
                type="button"
              />
            </div>
            <div
              aria-label="Modal"
              role="dialog"
              tabindex="-1"
            >
              <div>
                Goose
              </div>
            </div>
          </body>
        }
      >
        <div
          aria-label="Modal"
          role="dialog"
          tabIndex={-1}
        >
          <div>
            Goose
          </div>
        </div>
      </Portal>
    </ConditionalPortal>
  </ModalCurtain>
</div>
`;
